pytorch_fairseq_translation.md

```diff
@@ -0,0 +1,122 @@
+powerful sequence-to-sequence modeling architecture capable of producing
+state-of-the-art neural machine translation (NMT) systems.
+
+Recently, the fairseq team has explored large-scale semi-supervised training of
+Transformers using back-translated data, further improving translation quality
+over the original model. More details can be found in [this blog post][2].
+
+
+### Requirements
+
+We require a few additional Python dependencies for preprocessing:
+
+```bash
+pip install regex requests sacremoses sentencepiece subword_nmt
+```
+
+
+### English-to-French Translation
+
+To translate from English to French using the model from the paper [Scaling
+Neural Machine Translation][3]:
+
+```python
+import torch
+en2fr = torch.hub.load(
+  'pytorch/fairseq',
+  'transformer',
+  model_name_or_path='transformer.wmt14.en-fr',
+  data_name_or_path='.',
+  tokenizer='moses',
+  aggressive_dash_splits=True,
+  bpe='subword_nmt',
+)
+print(en2fr.generate('Hello world!'))
+# Bonjour à tous !
+```
+
+
+### English-to-German Translation
+
+Semi-supervised training with back-translation is an effective way of improving
+translation systems. In the paper [Understanding Back-Translation at Scale][4],
+we back-translate over 200 million German sentences to use as additional
+training data. An ensemble of these models was the winning submission to the
+[WMT'18 English-German news translation competition][5].
+
+To translate from English to German using one of the models from the winning
+ensemble (note: the archive includes the full ensemble and is 16GB uncompressed,
+so will take some time to download and extract):
+
+```python
+import torch
+en2de = torch.hub.load(
+  'pytorch/fairseq',
+  'transformer',
+  model_name_or_path='transformer.wmt18.en-de',
+  checkpoint_file='wmt18.model1.pt',
+  data_name_or_path='.',
+  tokenizer='moses',
+  aggressive_dash_splits=True,
+  bpe='subword_nmt',
+)
+len(en2de.models)
+# 1
+print(en2de.generate('Pytorch Hub is a pre-trained model repository designed to facilitate research reproducibility.'))
+# Pytorch Hub ist ein vortrainiertes Modelldepot, das die Reproduzierbarkeit von Forschungsergebnissen erleichtern soll.
+```
+
+We can also translate using the whole ensemble, but this requires a lot of GPU
+memory, so you may want to add the ``cpu=True`` flag below to generate on CPU
+instead:
+
+```python
+import torch
+en2de_ensemble = torch.hub.load(
+  'pytorch/fairseq',
+  'transformer',
+  model_name_or_path='transformer.wmt18.en-de',
+  checkpoint_file='wmt18.model1.pt:wmt18.model2.pt:wmt18.model3.pt:wmt18.model4.pt:wmt18.model5.pt',
+  data_name_or_path='.',
+  tokenizer='moses',
+  aggressive_dash_splits=True,
+  bpe='subword_nmt',
+)
+len(en2de_ensemble.models)
+# 5
+print(en2de_ensemble.generate('Pytorch Hub is a pre-trained model repository designed to facilitate research reproducibility.'))
+# Pytorch Hub ist ein vortrainiertes Modellarchiv, das die Reproduzierbarkeit der Forschung erleichtern soll.
+```
+
```

Comment:
Since we are planning to add more translation models in different languages in the future it might be good to show how to list the other available models or point them to the fairseq repo to find the rest of them.

Reply:
@nng555 In case it's helpful, users can find out all available hub models using `torch.hub.list(repo)`. Is it what you are looking for here?